### PR TITLE
Add the developer utility namemapdump.c

### PR DESCRIPTION
**Pull request by levitte**
This utility will display the namemap produced when fetching just
certain methods, or fetching them all, or none.  The latter will show
the namemap that's inherited from legacy methods.

### CODE REVIEW

**Comment by mspncp** [comment]
If it’s a command (line utility), why not name it like one? E.g. dump-namemap or dump_namemap?

**Comment by richsalz** [comment]
I think you should merge this into the standard app.  Perhaps as an option under the `list` command.  Then you get the benefit of being able to specify which providers are loaded, etc.  You could also add arguments such as cipher names, to trace specific behaviors.

**Comment by mspncp** [comment]
Sounds like a good idea.

**Comment by levitte** [comment]
> If it’s a command (line utility), why not name it like one? E.g. dump-namemap or dump_namemap?

So.... "tcpdump" isn't a command line utility?

**Comment by levitte** [comment]
> Then you get the benefit of being able to specify which providers are loaded, etc.  You could also add arguments such as cipher names, to trace specific behaviors.

Uhmmmmm, this program does all that...

File: util/namemapdump.c [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+#include <openssl/kdf.h>
+#include <openssl/store.h>
+#include <openssl/err.h>
+#include <openssl/provider.h>
+#include <openssl/safestack.h>
+#include "internal/namemap.h"
+
+static void dump_name(const char *name, void *data)
+{
+    int *first = data;
```
Comment (richsalz) [inline util/namemapdump.c:25]: did you mean this to be static?

**Comment by richsalz** [comment]
Why don't you want to bundle it in the openssl app?

**Comment by levitte** [comment]
> Why don't you want to bundle it in the openssl app?

I view it as a *user* app, not a *developer* app...  it's possible that I make too strict a difference.  But look at the discussions we've had re if "fipsinstall" should really be part of he the openssl app or not; I basically view "namemapdump" to be in the same category as "fipsinstall".

**Comment by levitte** [comment]
Side note: I wouldn't mind having a separate developer tool for OpenSSL...  `ossltool` has come to mind more than once

**Comment by richsalz** [comment]
> I basically view "namemapdump" to be in the same category as "fipsinstall"

Which is bundled :)  And so is `rehash`  Everything else in `util` is for *building* openssl.

As someone who developed a product (at a startup and then IBM acquired) for nearly a decade, I strongly encourage you to not make a distinction between dev tools and app tools.  (I remember a 6am Easter Sunday support call, where L3 support said "if I could run that command, you'd still be asleep" :)

I would expect app devs to say "why can't you find *my* special cipher, or *my* implementation of AES" and this tool will help with that.

**Comment by t-j-h** [comment]
If it is useful enough to want to have a program to run like this then the API should be public (not internal) and it should be in the openssl app).

It looks useful to me - although the program could do with being a bit more table driven rather than copied code with just function names changing - looks a lot cleaner if that was a table of opt+initfunc+cleanupfunc.

**Comment by levitte** [comment]
> If it is useful enough to want to have a program to run like this then the API should be public (not internal) and it should be in the openssl app).

And therein lies another problem.  The namemap is entirely internal, which is intentional.  I forgot for a moment, but that was another reason to have it as a dev program for internal use.

**Comment by mspncp** [comment]
> > If it’s a command (line utility), why not name it like one? E.g. dump-namemap or dump_namemap?
> 
> So.... "tcpdump" isn't a command line utility?

You beat me. ;-)

**Review by slontis — COMMENTED** [review]
Why didnt you just make this an app?

**Comment by t8m** [comment]
I do not think this belongs to openssl app. This seems to me like an low-level developer tool.

**Comment by levitte** [comment]
I'm withdrawing this

**Pull request closed without merging**